# QA — Trial base sign-up + scheduling improvements (PR #197)

Branch: `feat/trial-base-signup-scheduling` (head `e3b0b58`)

## Gate (AC20)

`npx tsc --noEmit` output on the branch:

```
src/interactions/moderation/TrialReport.ts(2,47): error TS6133: 'ModalSubmitInteraction' ...
src/modules/LeaderboardHandler.ts(65,27): error TS2339: Property 'setYear' ...
src/modules/ReportHandler.ts(1,150): error TS6133: 'Message' ...
src/modules/ReportHandler.ts(1,211): error TS6133: 'RoleSelectMenuBuilder' ...
src/modules/UtilityHandler.ts(3,25): error TS2307: Cannot find module '../../config.json' ...
```

All 5 errors are PRE-EXISTING on the `production` baseline and are in files this PR does **not**
touch. The 4 changed files (`ScheduledTrial.ts`, `GuildSpecifics.ts`, `ScheduledTrialHandler.ts`,
`ReminderHandler.ts`) introduce **ZERO new errors**. Gate: **PASS for this PR** (diff is type-clean).

## Code trace (AC -> evidence)

| AC | Status | Evidence (file:line) |
|----|--------|----------------------|
| 1  | PASS | `ScheduledTrialHandler.ts:108-110` Base line, `<@id>` / `_None yet_`, only when `!isTicket` |
| 2  | PASS | `ScheduledTrialHandler.ts:430-433` eligibility = trialees OR fills |
| 3  | PASS | base is a flag (`baseId`), no slot added; `maxFills` (`:181-183`) and `maxTrialees` untouched by base |
| 4  | PASS | `ScheduledTrialHandler.ts:424-427` reject naming `<@baseId>`, no state change |
| 5  | PASS | `ScheduledTrialHandler.ts:416-422` toggle off -> `baseId=null` |
| 6  | PASS | `ScheduledTrialHandler.ts:430-433` not-signed-up rejected |
| 7  | PASS | trialee withdraw `:345`, fill withdraw `:400`, manager remove `:592` all clear `baseId` |
| 8  | PASS | `ScheduledTrial.ts:45-47` `@Column text nullable baseId`; persisted via `repo.save` |
| 9  | PASS | `ReminderHandler.ts:62` `targetChannelId = channelIds.trialeeChat || trial.channelId`; MAIN has `trialeeChat=1468293993258553355` (`GuildSpecifics.ts:115`) |
| 10 | PASS | TEST guild block has no `trialeeChat` -> `undefined` -> falls back to `trial.channelId` |
| 11 | PASS | `ReminderHandler.ts:63-68` fetch `.catch(()=>null)`, fall back to `trial.channelId`; reminderSent still set downstream |
| 12 | PASS | `ScheduledTrialHandler.ts:297-301` pings `roleIds.trialTeam` + `roleIds[getTrialeeRoleKey(tier)]` |
| 13 | PASS | `ScheduledTrialHandler.ts:303` `pingChannelId = channelIds.trialeeChat || channel.id` |
| 14 | PASS | `ScheduledTrialHandler.ts:313` `allowedMentions: { roles: pingRoleIds }` (no parse/everyone) |
| 15 | PASS | `ScheduledTrialHandler.ts:301` filters falsy + sentinel `000000000000000000`; sends with whatever resolves; wrapped in try/catch (`:316`) so scheduling never aborts |
| 16 | PASS | `ScheduledTrialHandler.ts:181-183` `Math.max(0, 5 - 1 - minTrialees)` |
| 17 | PASS | `ScheduledTrialHandler.ts:422-426` rejects at derived cap with max in message |
| 18 | PASS | `ScheduledTrialHandler.ts:108` card label uses `maxFills(trial)` |
| 19 | PASS | `totalPlayers` cap (`:107`/`MAX_PARTICIPANTS`) and `maxTrialees` enforcement unchanged; base never blocks |
| 20 | PASS | see Gate |
| 21 | PASS | no `permissions` getter / role-string changes in diff; base reuses "must be signed up" gate |

No QA CRITICAL gaps found — every AC has an implementation.

---

## Manual Discord test checklist

Preconditions for all tests: a `/schedule-trial` card posted in the scheduling channel. To schedule
you need the host/trial-team permission the command already requires. Use the TEST guild for most
checks; AC9/AC12 channel routing needs the MAIN guild (`885457551397912596`).

Glossary of test actors:
- **TRIALEE-A / TRIALEE-B**: users holding the tier's trialee role.
- **FILL-A**: user with `trialTeam`/admin/owner (eligible to fill).
- **OUTSIDER**: user with no sign-up on the card.

### Group A — Base sign-up (AC1-8)

1. **(AC1) Base line renders.** Schedule a non-ticket trial. Open the card.
   Expect: a `**Base:** _None yet_` line, plus a second action row with a `Base sign up` button.
   (Ticket trials must NOT show this line/button.)

2. **(AC2, AC6) Outsider rejected.** As OUTSIDER (not signed up), click `Base sign up`.
   Expect ephemeral: "You must be signed up as a trialee or fill before you can sign up as the base."
   Card unchanged (`_None yet_`).

3. **(AC2) Trialee claims base.** As TRIALEE-A click trialee sign-up, then click `Base sign up`.
   Expect ephemeral "You are now the base for this trial." Card `**Base:** @TRIALEE-A`. Button label
   now reads `Base sign up / step down`.

4. **(AC4) Second user rejected, holder named.** As FILL-A (signed up as fill), click `Base sign up`
   while TRIALEE-A holds base.
   Expect ephemeral: "A base is already assigned to @TRIALEE-A." Card still shows @TRIALEE-A.

5. **(AC2) Fill can claim (after release).** Have TRIALEE-A release (test 6) then as FILL-A click
   `Base sign up`. Expect card `**Base:** @FILL-A` — confirms fills are eligible.

6. **(AC5) Holder toggles off.** As the current base holder click `Base sign up` again.
   Expect ephemeral "You are no longer the base for this trial." Card returns to `_None yet_`.

7. **(AC7) Base auto-clears on trialee withdraw.** TRIALEE-A is base; TRIALEE-A clicks the trialee
   sign-up button again (withdraw). Expect "You have withdrawn from this trial." AND card Base line
   back to `_None yet_` (TRIALEE-A removed from trialees too).

8. **(AC7) Base auto-clears on fill withdraw.** Same as 7 but with FILL-A as base withdrawing via the
   fill button. Card Base -> `_None yet_`.

9. **(AC7) Base auto-clears on manager removal.** Base = TRIALEE-A. A host/trial-team manager uses the
   remove control and selects TRIALEE-A. Expect TRIALEE-A gone from trialees AND Base -> `_None yet_`.

10. **(AC8) Persistence.** With a base set, force a card re-render (any sign-up action by another user,
    or restart the bot). Expect the Base line still shows the same `@holder` (read back from DB).

11. **(AC3) Base adds no slot.** With base set, check `_Total players: N/5_` did NOT increase when base
    was flagged, and trialee/fill counts are unchanged by flagging base.

### Group B — Fill cap (AC16-19)

12. **(AC16, AC18) minTrialees=1 -> max fills 3.** Schedule with minTrialees=1.
    Expect card line `Trial Team Fills (0/3)`.

13. **(AC18) minTrialees=4 -> max fills 0.** Schedule with minTrialees=4.
    Expect card line `Trial Team Fills (0/0)`.

14. **(AC17) Fill rejected at cap.** On the minTrialees=4 (cap 0) trial, click fill sign-up.
    Expect ephemeral: "All fill slots are already taken (max 0)." On a cap-3 trial, sign up 3 fills then
    a 4th — 4th gets "(max 3)".

15. **(AC19) Total/trialee caps unchanged.** Fill the team toward 5 total (host + trialees + fills) and
    confirm the existing total-player cap (5 incl. host) and `maxTrialees` still block as before.
    Flagging a base must NOT block any further sign-up or the finish flow.

### Group C — Reminder channel (AC9-11)

16. **(AC9) MAIN reminder -> trialee-chat.** On MAIN guild, schedule a trial ~31 min out with at least
    one trialee/fill. When the 30-min window hits, expect the reminder posted in channel
    `1468293993258553355` (trialee-chat), NOT the scheduling channel. To simulate without waiting: set
    `scheduledTime` in the DB to ~29 min from now (within the 30-min window) and wait for the
    ReminderHandler tick; the message should appear in trialee-chat and `reminderSent` flips true.

17. **(AC10) TEST reminder -> own channel.** Repeat on the TEST guild (no `trialeeChat` configured).
    Expect the reminder in the trial's own `channelId` exactly as before.

18. **(AC11) Fallback when trialee-chat unfetchable.** On MAIN, temporarily make trialee-chat
    unreachable (e.g. bot lacks view perms there). Expect the reminder to fall back to the trial's
    `channelId`, post there, and still mark `reminderSent` — no crash, no silent drop.

### Group D — Schedule-time ping (AC12-15)

19. **(AC12, AC14) Ping fires with both roles.** Schedule a trial for a tier. Expect a message in
    trialee-chat (MAIN) / scheduling channel (TEST) that mentions BOTH the global trial-team role AND
    that tier's trialee role, and that the message links back to the card channel. Confirm the two roles
    actually trigger a notification (members of those roles get pinged) and NOTHING else is mass-pinged
    (no @everyone/@here, no unrelated roles).

20. **(AC12) Tier variance.** Schedule different tiers; expect the trial-team role to be the same every
    time but the trialee role to change with the tier.

21. **(AC13) TEST channel self-contained.** On TEST (no trialeeChat), confirm the ping lands in the same
    scheduling channel as the card.

22. **(AC15) Missing/sentinel role id.** Schedule a tier whose trialee role id is missing/unconfigured
    (or the sentinel `000000000000000000`). Expect scheduling to still succeed and the card to post; the
    ping sends with only the resolvable role(s) — it must NOT throw or abort scheduling. If BOTH ids are
    unresolvable, no ping is sent but scheduling still completes.

### Group E — Non-regression (AC19, AC21)

23. **(AC21) Permissions unchanged.** Confirm: trialee sign-up still requires the tier trialee role;
    fill sign-up still requires trialTeam/admin/owner; remove/manage still requires host-or-trialteam;
    base claim adds no new permission (only "must be signed up"). An OUTSIDER cannot manage or remove.

24. **(AC19) Finish flow unaffected by base.** With a base flagged, run the finish flow and confirm it
    behaves identically to before (base flag is purely informational).

## Notes / risks
- AC9/AC12/AC13 channel routing depends on `client.channelIds.trialeeChat` resolving correctly per
  guild. Verified: MAIN block sets `trialeeChat=1468293993258553355`; TEST block omits it (-> undefined
  -> fallback). Confirm in Discord that the bot can actually view/send in trialee-chat on MAIN.
- The 30-min reminder cannot be triggered on demand via Discord UI; tests 16-18 require either waiting
  or editing `scheduledTime`/`reminderSent` in the sqlite DB to land inside the window.

--- a/src/GuildSpecifics.ts
+++ b/src/GuildSpecifics.ts
@@ -55,7 +55,7 @@ export function getChannels(guildId: string | undefined): Channels {
             masterTrialee1000TicketsCategory: '1492216896852988026',
             masterTrialee2000TicketsCategory: '1492216981972058122',
             grandmasterTrialee2000TicketsCategory: '1492216981972058122',
-            trialScheduling: '1391860639584096316'
+            trialScheduling: '1495724268041011383'
         }
     }
 

--- a/src/modules/ScheduledTrialHandler.ts
+++ b/src/modules/ScheduledTrialHandler.ts
@@ -468,9 +468,9 @@ export default class ScheduledTrialHandler {
             return;
         }
 
-        // Must already be a signed-up participant (trialee or fill) to claim base.
-        if (!trial.trialees.includes(member.id) && !trial.fills.includes(member.id)) {
-            await interaction.editReply('You must be signed up as a trialee or fill before you can sign up as the base.');
+        // Must be the host or an already-signed-up participant (trialee or fill) to claim base.
+        if (member.id !== trial.hostId && !trial.trialees.includes(member.id) && !trial.fills.includes(member.id)) {
+            await interaction.editReply('You must be the host or signed up as a trialee or fill before you can sign up as the base.');
             return;
         }
 

